File: src/main/java/io/github/mavenindexdoctor/application/MavenDiagnosticsService.java
package io.github.mavenindexdoctor.application;

import com.intellij.openapi.project.Project;
import io.github.mavenindexdoctor.domain.ArtifactCoordinate;
import io.github.mavenindexdoctor.domain.DiagnosticIssue;
import io.github.mavenindexdoctor.domain.DiagnosticIssueType;
import io.github.mavenindexdoctor.domain.DiagnosticSeverity;
import io.github.mavenindexdoctor.domain.MavenDiagnosticReport;
import org.jetbrains.idea.maven.indices.MavenIndicesManager;
import org.jetbrains.idea.maven.project.MavenGeneralSettings;
import org.jetbrains.idea.maven.project.MavenProjectsManager;
import org.jetbrains.idea.maven.utils.MavenUtil;

import java.io.IOException;
import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class MavenDiagnosticsService {

    private static final String DEFAULT_SETTINGS_DIRECTORY = ".m2";
    private static final String DEFAULT_SETTINGS_FILE = "settings.xml";
    private static final String INDEX_DIRECTORY = "Indices";

    private final Project project;
    private final MavenFileScanner fileScanner;
    private final MavenSettingsScanner settingsScanner;

    public MavenDiagnosticsService(Project project) {
        this.project = project;
        this.fileScanner = new MavenFileScanner();
        this.settingsScanner = new MavenSettingsScanner();
    }

    public MavenDiagnosticReport scan() throws IOException {
        MavenProjectsManager projectsManager = MavenProjectsManager.getInstance(project);
        Path repositoryPath = projectsManager.getRepositoryPath();
        Path indexPath = MavenUtil.getPluginSystemDir(INDEX_DIRECTORY);
        Path settingsPath = resolveSettingsPath(projectsManager.getGeneralSettings());

        List<DiagnosticIssue> issues = new ArrayList<>();
        issues.addAll(fileScanner.scanLastUpdatedFiles(repositoryPath));
        issues.addAll(fileScanner.scanIndexDirectory(indexPath));
        issues.addAll(settingsScanner.scan(settingsPath));
        inspectIdeaMavenSettings(projectsManager.getGeneralSettings(), settingsPath, issues);

        List<ArtifactCoordinate> artifacts = fileScanner.findLocalArtifacts(repositoryPath);
        int missingFromIndexCount = compareRepositoryWithIndex(artifacts, issues);
        issues.sort(Comparator.comparingInt(issue -> severityRank(issue.severity())));

        return new MavenDiagnosticReport(
                Instant.now(),
                repositoryPath,
                indexPath,
                settingsPath,
                artifacts.size(),
                missingFromIndexCount,
                List.copyOf(issues)
        );
    }

    public Path resolveSettingsPath() {
        return resolveSettingsPath(MavenProjectsManager.getInstance(project).getGeneralSettings());
    }

    private int compareRepositoryWithIndex(
            List<ArtifactCoordinate> artifacts,
            List<DiagnosticIssue> issues
    ) {
        MavenIndicesManager indicesManager = MavenIndicesManager.getInstance(project);
        indicesManager.scheduleUpdateIndicesListAndWait();
        indicesManager.waitForGavUpdateCompleted();

        int missingCount = 0;
        for (ArtifactCoordinate artifact : artifacts) {
            if (indicesManager.hasLocalVersion(
                    artifact.groupId(),
                    artifact.artifactId(),
                    artifact.version()
            )) {
                continue;
            }
            missingCount++;
            issues.add(new DiagnosticIssue(
                    DiagnosticIssueType.ARTIFACT_MISSING_FROM_INDEX,
                    DiagnosticSeverity.ERROR,
                    "Artifact missing from IDEA index",
                    artifact + " exists in the local repository but is absent from IDEA's local Maven index.",
                    null,
                    "Refresh the local repository index."
            ));
        }
        return missingCount;
    }

    private Path resolveSettingsPath(MavenGeneralSettings settings) {
        String configuredSettings = settings.getUserSettingsFile();
        if (configuredSettings != null && !configuredSettings.isBlank()) {
            return Path.of(configuredSettings).toAbsolutePath().normalize();
        }
        return Path.of(
                System.getProperty("user.home"),
                DEFAULT_SETTINGS_DIRECTORY,
                DEFAULT_SETTINGS_FILE
        ).toAbsolutePath().normalize();
    }

    private void inspectIdeaMavenSettings(
            MavenGeneralSettings settings,
            Path settingsPath,
            List<DiagnosticIssue> issues
    ) {
        if (settings.isWorkOffline()) {
            issues.add(new DiagnosticIssue(
                    DiagnosticIssueType.MAVEN_OFFLINE_MODE,
                    DiagnosticSeverity.ERROR,
                    "IDEA Maven offline mode is enabled",
                    "IDEA cannot download dependencies or repository index updates while Maven is offline.",
                    settingsPath,
                    "Disable Work offline in IDEA Maven settings and refresh indexes."
            ));
        }
    }

    private int severityRank(DiagnosticSeverity severity) {
        return switch (severity) {
            case ERROR -> 0;
            case WARNING -> 1;
            case INFO -> 2;
        };
    }
}

File: src/main/kotlin/io/github/mavenindexdoctor/application/MavenDiagnosticsService.kt
package io.github.mavenindexdoctor.application

import com.intellij.openapi.application.PathManager
import io.github.mavenindexdoctor.domain.DiagnosticIssue
import io.github.mavenindexdoctor.domain.DiagnosticSeverity
import io.github.mavenindexdoctor.domain.MavenDiagnosticReport
import java.nio.file.Files
import java.nio.file.Path
import java.time.Instant

class MavenDiagnosticsService {
    fun scan(): MavenDiagnosticReport {
        val repositoryPath = Path.of(System.getProperty("user.home"), ".m2", "repository")
        val indexPath = Path.of(PathManager.getSystemPath(), "maven", "indices")
        val issues = buildList {
            addAll(scanLastUpdatedFiles(repositoryPath))
            addAll(scanIndexDirectory(indexPath))
        }

        return MavenDiagnosticReport(Instant.now(), repositoryPath, indexPath, issues)
    }

    private fun scanLastUpdatedFiles(repositoryPath: Path): List<DiagnosticIssue> {
        if (!Files.exists(repositoryPath)) {
            return listOf(
                DiagnosticIssue(
                    DiagnosticSeverity.WARNING,
                    "Local Maven repository not found",
                    "The default local repository does not exist yet.",
                    repositoryPath,
                ),
            )
        }

        Files.walk(repositoryPath).use { paths ->
            return paths
                .filter { Files.isRegularFile(it) && it.fileName.toString().endsWith(".lastUpdated") }
                .map {
                    DiagnosticIssue(
                        DiagnosticSeverity.WARNING,
                        "Failed Maven download marker",
                        "Maven left a .lastUpdated marker; the dependency may be skipped until it is retried.",
                        it,
                    )
                }
                .toList()
        }
    }

    private fun scanIndexDirectory(indexPath: Path): List<DiagnosticIssue> {
        if (!Files.exists(indexPath)) {
            return listOf(
                DiagnosticIssue(
                    DiagnosticSeverity.INFO,
                    "Maven index directory not found",
                    "The directory will be created when IDEA downloads its first Maven index.",
                    indexPath,
                ),
            )
        }
        if (!Files.isReadable(indexPath)) {
            return listOf(
                DiagnosticIssue(
                    DiagnosticSeverity.ERROR,
                    "Maven index directory is not readable",
                    "IDEA cannot inspect the Maven index directory.",
                    indexPath,
                ),
            )
        }

        Files.walk(indexPath).use { paths ->
            return paths
                .filter { Files.isRegularFile(it) && Files.size(it) == 0L }
                .map {
                    DiagnosticIssue(
                        DiagnosticSeverity.ERROR,
                        "Empty Maven index file",
                        "A zero-byte index file may indicate an interrupted or corrupted download.",
                        it,
                    )
                }
                .toList()
        }
    }
}
